client/lib/registration.js

```diff
@@ -0,0 +1,257 @@
+            resolve(res);
+          }
+        }).catch((res) => {
+          logger.error('Failed to register:', res);
+          reject(res);
+        });
+      }
+    });
+  }
+
+  /*
+   * Generate the keys necessary for this instance
+   *
+   * @return Boolean
+   */
+  generateKeys() {
+    let ecdh = crypto.createECDH('secp521r1');
+    let pubKey = ecdh.generateKeys('base64');
+    logger.info('Generated public key:', pubKey);
+
+    this.publicKey = pubKey;
+    this.privateKey = ecdh.getPrivateKey('base64');
+
+    /* If we have a private key, that's good enough! */
+    if (this.privateKey) {
+      return true;
+    }
+
+    return false;
+  }
+
+  /*
+   * Return the base64 encoded public key associated with this instance
+   *
+   * Will return null if no public key has yet been generated
+   *
+   * @return String
+   */
+  getPublicKey() {
+    return this.publicKey;
+  }
+
+  /*
+   * Persist the UUID generated by the call to the registration service
+   *
+   * @return Boolean on success
+   */
+  saveUUIDSync() {
+    if (this.uuid != null) {
+      const config = {
+        uuid: this.uuid,
+      };
+      logger.info('Saving uuid to disk at', this.uuidPath());
+
+      fs.writeFileSync(
+        this.uuidPath(),
+        JSON.stringify(config),
+        this.fileOptions);
+      return true;
+    }
+    return false;
+  }
+
+  /*
+   * Persist generated keys
+   *
+   * @return Boolean on success
+   */
+  saveKeysSync() {
+    if ((!this.publicKey) || (!this.privateKey)) {
+      logger.warn('saveKeysSync() called without a private or public key on Registration');
+      return false;
+    }
+
+    const keyPath = this.keyPath();
+    const publicKeyPath = this.publicKeyPath();
+    const privateKeyPath = this.privateKeyPath();
+
+    logger.debug('Writing public key to', publicKeyPath);
+    fs.writeFileSync(publicKeyPath, this.publicKey, this.fileOptions);
+
+    logger.debug('Writing private key to', privateKeyPath);
+    fs.writeFileSync(privateKeyPath, this.privateKey, this.fileOptions);
+
+    return true;
+  }
+
+  /*
+   * Load generated keys from disk assuming they are present
+   *
+   * @return Boolean on successful load of keys
+   */
+  loadKeysSync() {
+    if ((this.publicKey != null) || (this.privateKey != null)) {
+      logger.warn('loadKeysSync() called despite public/private keys already having been loaded', this.publicKey);
+      return false;
+    }
+
+    if (!this.hasKeys()) {
+      return false;
+    }
+    this.publicKey = fs.readFileSync(this.publicKeyPath(), this.fileOptions);
+    this.privateKey = fs.readFileSync(this.privateKeyPath(), this.fileOptions);
+    return true;
+  }
+
+
+  /*
+   * Returns the default home directory or the value of EVERGREEN_HOME
+   */
+  homeDirectory() {
+    /* The default home directory is /evergreen, see the Dockerfile in the root
+     * directory of th repository
+     */
+    if (!process.env.EVERGREEN_HOME) {
+      return '/evergreen';
```

Comment:
Shouldn't we rather throw an exception? Or do you use that I guess when developing? Maybe it should log a warning ?